### DISCUSSION TITLE
GitHub Actions to create Xcode release

### DISCUSSION THREAD

**Discussion by AlexPerathoner** [sparkle-project/Sparkle#2308] [Show and tell]
# GitHub Actions + Xcode macOS app + Sparkle

First things first: I'm **lazy**. And the entire process of creating a release with Sparkle is a bit too much for me.

So better spend 20x times more time creating a GitHub Action that automates that: https://github.com/AlexPerathoner/SparkleReleaseTest

## What does it do?

The owner creates a PR to `master` and comments `/release`. If the PR is mergeable, the comment triggers a GitHub Action which will:
1. Extrapolate from `Release_Notes.md` the:
    * latest version number
    * its release notes
    * its title
    * previous version number
1. Replace all occurences of the previous version number with the new version number within the Xcode project
1. Archive and export the Xcode project as a .app
4. Run the `generate_appcast` tool from Sparkle, which will create/update the appcast.xml file in the Docs/Support folder
5. Create a GitHub release with the zipped app, the release notes, the latest version and the title
6. Merge the PR, which triggers the GitHub action to build GitHub Pages, which will host the `appcast.xml` file. The update is so available for all users.

---

Hope some of you can find this useful and save some time!

**Comment by zorgiepoo** [comment]
Nice idea and glad you were able to cook up something.

Some feedback upon quickly looking at it:

* I think the default branch name of GitHub repositories is `main` these days rather than `master`. An additional enhancement, if you have the need for it, might be to customize the target branch (for example I released Sparkle 2.3.1 on a separate branch while still working on 2.4.0 in our main branch (we have a GitHub action here for reference when making Sparkle releases)).
* It is better to pass "-" as the standard input file when using `--ed-key-file` in this situation; you don't want to save the private key file to disk. So: `echo "$PRIVATE_SPARKLE_KEY" | ./generate_appcast --ed-key-file -` (this is mentioned in the help page for generate_appcast)
* I believe this generates a release using only the latest built release archive, which means generate_appcast in this case cannot build/upload delta updates or preserve other update items (for example if two releases have different minimum OS requirements). You may be able to download previous releases to put into `Release/` using GitHub's APIs perhaps.
* An additional enhancement following the above point (that you may not need) may be to make pre-releases and using a Sparkle channel for beta updates.

Relatedly: #648
